vllm/ex/fusion.py

```diff
@@ -0,0 +1,284 @@
+    if not is_call(node):
+        return False
+
+    op_name = node_function_target(node)
+    if node.op == 'call_function':
+        return op_name in FUSABLE and not FUSABLE[op_name]
+    else:
+        # TODO: check class type
+        class_type = call_method_class(node)
+        return op_name in FUSABLE and not FUSABLE[op_name]
+
+"""
+Determine whether or not node is a fusable compute operation, e.g. gemm.
+"""
+def is_compute(node: torch.fx.Node) -> bool:
+    if not is_call(node):
+        return False
+
+    op_name = node_function_target(node)
+    if node.op == 'call_function':
+        return op_name in FUSABLE and FUSABLE[op_name]
+    else:
+        # TODO: check class type
+        class_type = call_method_class(node)
+        return op_name in FUSABLE and FUSABLE[op_name]
+
+
+def is_getitem(a: torch.fx.Node) -> bool:
+    return is_call(a) and node_function_target(a) == '_operator.getitem'
+
+
+"""
+Are nodes a and b fusable together?
+This function assumes 'b' is a direct successor of 'a'.
+"""
+def is_fusable_pair(a: torch.fx.Node, b: torch.fx.Node) -> bool:
+    return is_fusable(a) and is_fusable(b)
+
+
+"""
+Are nodes 'a' and 'b' fusable together and is 'a' optionally a compute op?
+This function assumes 'b' is a direct successor of 'a'.
+"""
+def is_compute_fusable_pair(a: torch.fx.Node, b: torch.fx.Node) -> bool:
+    return (is_fusable(a) or is_compute(a)) and is_fusable(b)
+
+
+"""
+Determine if any kwargs associated with 'node' are supported.
+"""
+def supported_kwargs(node: torch.fx.Node, allow_const_kwargs: bool = False) -> bool:
+    if allow_const_kwargs:
+        for arg in node.kwargs.values():
+            if not isinstance(arg, torch.fx.node.BaseArgumentTypes):
+                return False
+        return True
+    else:
+        return node.kwargs is None or len(node.kwargs) == 0
+
+
+"""
+1. create Partition objects from sequences of fusable nodes
+2. use fuse_partitions to recreate the graph torch._inductor.fx_passes.group_batch_fusion
+"""
+def pointwise_fusion(
+    cc: CodeCache,
+    fgen: FusedOpGenerator,
+    mod: torch.fx.GraphModule,
+    example_inputs: List[torch.Tensor],
+    fuse_inputs: bool = False,
+    fuse_with_compute=True
+) -> torch.fx.GraphModule:
+    # find all groups of nodes that can be fused and assign to
+    # unique partition id, i.e. map_node
+
+    fg = FlowGraph(mod)
+
+    ShapeProp(mod).propagate(*example_inputs)
+
+    node_map = dict()
+    partition = 0
+
+    def map_node(n: torch.fx.Node) -> int:
+        return node_map[n]
+
+    # assumption, graph.nodes are in topo order
+    mod.graph.lint()
+
+    logger.debug("start fusion")
+
+    # create partition groups
+    # run in reverse order so predecesors of non-unary ops will appear
+    # in the same partition.
+    for n in reversed(mod.graph.nodes):
+        logger.debug(f"CONSIDER {n}")
+
+        if not is_call(n):
+            logger.debug(f"  REJECT {n} not call")
+            node_map[n] = 0
```

Comment:
I think it would be good to add a name for the `0` partition number. Maybe something like `UNFUSED_PARTITION`?